core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java

```diff
@@ -37,98 +33,78 @@
  */
 abstract class AbstractEventContext implements BaseEventContext {
 
+  final static int STATE_READY = 0;
+  final static int STATE_DONE = 1;
+  final static int STATE_COMPLETE = 2;
+  final static int STATE_TERMINATED = 3;
+
   private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventContext.class);
   protected static final FlowExceptionHandler NULL_EXCEPTION_HANDLER = NullExceptionHandler.getInstance();
 
-  private transient MonoProcessor<CoreEvent> beforeResponseProcessor;
-  private transient MonoProcessor<CoreEvent> responseProcessor;
-  private transient MonoProcessor<Void> completionProcessor;
-  private transient Disposable completionSubscriberDisposable;
   private transient final List<BaseEventContext> childContexts = new LinkedList<>();
-  private transient Mono<Void> completionCallback;
-  private transient FlowExceptionHandler exceptionHandler;
+  private transient final FlowExceptionHandler exceptionHandler;
+  private final MonoProcessor<CoreEvent> responseProcessor = MonoProcessor.create();
+  private volatile int state = STATE_READY;
+  private volatile boolean externalCompletion = true;
+  private final List<BiConsumer<CoreEvent, Throwable>> onResponseConsumerList = new ArrayList<>();
+  private final List<BiConsumer<CoreEvent, Throwable>> onCompletionConsumerList = new ArrayList<>();
+  private final List<BiConsumer<CoreEvent, Throwable>> onTerminatedConsumerList = new ArrayList<>();
 
   public AbstractEventContext() {
-    this(NULL_EXCEPTION_HANDLER, empty());
+    this(NULL_EXCEPTION_HANDLER, null);
   }
 
   public AbstractEventContext(FlowExceptionHandler exceptionHandler) {
-    this(exceptionHandler, empty());
+    this(exceptionHandler, null);
   }
 
   public AbstractEventContext(FlowExceptionHandler exceptionHandler, Publisher<Void> completionCallback) {
-    this.completionCallback = from(completionCallback);
+    if (completionCallback != null) {
+      externalCompletion = false;
+      from(completionCallback).doOnSuccessOrError((aVoid, throwable) -> {
+        externalCompletion = true;
+        tryTerminate();
+      }).subscribe();
```

Comment:
is it correct to subscribe this here? woudln't that execute it right now on this thread?

Reply:
It registers interest in this thread, which is fine.